### DISCUSSION TITLE
Could not create the driver from NHibernate.Driver.SqlClientDriver

### DISCUSSION THREAD

**Discussion by chopswil15** [nhibernate/nhibernate-core#2661] [Q&A] [answered]
I'm moving from .Net Framework to Core 3.1
the below code works in .net framework but in Core i get an error.
The System.Data.SqlClient dll is in the bin folder and has PackageReference Include="Microsoft.Data.SqlClient" Version="2.0.0"  in the project.

I'm not understanding what the issue is here.
Thoughts?

there is this message
HibernateException: The DbCommand and DbConnection implementation in the assembly System.Data.SqlClient could not be found. Ensure that the assembly System.Data.SqlClient is located in the application directory or in the Global Assembly Cache. If the assembly is in the GAC, use <qualifyAssembly/> element in the application configuration file to specify the full name of the assembly.

**The Error:**
```
DatabaseInterfaces.DatabaseLibraryException
  HResult=0x80131500
  Message=: Could not create the driver from NHibernate.Driver.SqlClientDriver, NHibernate, Version=5.3.0.0, Culture=neutral, PublicKeyToken=aa95f207798dfdb4.
  Source=NHibernateLibraries
  StackTrace:
   at NHibernateLibraries.NHibernateLibrarySession..ctor(String connectionString, String DomainAssemblyName) in C:\Users\Owner\Documents\GitHub\PathfinderControlLibrariesCore\NHibernateLibraries\NHibernateLibrarySession.cs:line 32
   at NHibernateLibraries.NHibernateRepository`1..ctor(String connectionString, String domainAssemblyName) in C:\Users\Owner\Documents\GitHub\PathfinderControlLibrariesCore\NHibernateLibraries\NHibernateRepository.cs:line 21
   at PathfinderContext.NHibernateServiceBase.FindBy[T](Int32 id) in 
```
**The Code:**
```csharp
   private ISessionFactory CreateSessionFactory(string assemblyNameWithClassMapper)
        {
            var assembly = Assembly.Load(assemblyNameWithClassMapper);
            if (assembly == null) throw new System.Exception(assemblyNameWithClassMapper + " referance is null");
            var configuartion = Fluently.Configure()
                .Database(MsSqlConfiguration.MsSql2012.ConnectionString(_connectionString))
                .Mappings(m => m.FluentMappings.AddFromAssembly(assembly));            
            return configuartion.BuildSessionFactory();
        }
```

**Comment by hazzik** [comment, marked as answer]
`Microsoft.Data.SqlClient` requires `NHibernate.Driver.MicrosoftDataSqlClientDriver`

**Reply by chopswil15** [reply]
dumb question, where do i get NHibernate.Driver.MicrosoftDataSqlClientDriver?
I searched for a nuget packed like this name but could not find one.

**Reply by chopswil15** [reply]
in the Nhibernate.xml that is installed when i installed NHibernate 5.3.5 it has this entry
doesn't this mean that NHibernate.Driver.MicrosoftDataSqlClientDriver is part of NHibernate.dll?
  
member name="T:NHibernate.Driver.MicrosoftDataSqlClientDriver">
            summary
                A NHibernate Driver for using the SqlClient DataProvider
            /summary
        /member

**Reply by chopswil15** [reply]
I got it, i added this to my web api project file and everything works!!
PackageReference Include="System.Data.SqlClient" Version="4.8.2" /

**Reply by chopswil15** [reply]
So i have a separate assembly that i call NHibernateLibraries that has all my code for creating and managing a session. 
I have another assembly that uses my domain objects and contains the fluent mappers of those domain objects.
It seems when i have when web api project which imports these assemblies it doesn't know about the includes of these other assemblies and have to be manually added to that project.

**Reply by oskarb** [reply]
NHibernate loads drivers by reflection - otherwise you might need to have a bunch of drivers for database products you are not even using present. So build process are likely to not pick up on the dependencies automatically.

**Reply by hazzik** [reply]
The package `System.Data.SqlClient` is required `NHibernate.Driver.SqlClientDriver` driver and package `Microsoft.Data.SqlClient` requires `NHibernate.Driver.MicrosoftDataSqlClientDriver`. If you want to use `Microsoft.Data.SqlClient` you would need to switch your driver settings. How your code is organized does not really matter.
